daemon/src/engine/client/cl_parse.cpp

```diff
@@ -108,148 +84,83 @@ CL_ParsePacketEntities
-			if ( cl_shownet->integer == 3 )
-			{
-				Com_Printf( "%3i:  unchanged: %i\n", msg->readcount, oldnum );
-			}
-
-			CL_DeltaEntity( msg, newframe, oldnum, oldstate, true );
-
-			oldindex++;
-
-			if ( oldindex >= oldframe->numEntities )
-			{
-				oldnum = MAX_GENTITIES;
-			}
-			else
-			{
-				oldstate = &cl.parseEntities[
-				             ( oldframe->parseEntitiesNum + oldindex ) & ( MAX_PARSE_ENTITIES - 1 ) ];
-				oldnum = oldstate->number;
-			}
-		}
-
-		if ( oldnum == newnum )
-		{
-			// delta from previous state
-			if ( cl_shownet->integer == 3 )
-			{
-				Com_Printf( "%3i:  delta: %i\n", msg->readcount, newnum );
-			}
-
-			CL_DeltaEntity( msg, newframe, newnum, oldstate, false );
-
-			oldindex++;
-
-			if ( oldindex >= oldframe->numEntities )
-			{
-				oldnum = MAX_GENTITIES;
-			}
-			else
-			{
-				oldstate = &cl.parseEntities[
-				             ( oldframe->parseEntitiesNum + oldindex ) & ( MAX_PARSE_ENTITIES - 1 ) ];
-				oldnum = oldstate->number;
-			}
+    // The entity packet contains the delta between the two snapshots, with data only
+    // for entities that were created, changed or removed. Entities entries are in
+    // order of increasing entity number, as are entities in a snapshot. Using this we
+    // have an efficient algorithm to create the new snapshot, that goes over the old
+    // snapshot once from the beginning to the end.
+
+    // If we don't have an old snapshot or it is empty, we'll recreate all entities
+    // from the baseline entities as setting oldEntityNum to MAX_GENTITIES will force
+    // us to only do step (3) below.
+    unsigned int oldEntityNum = MAX_GENTITIES;
+    if (oldSnapshot && oldSnapshot->entities.size() > 0){
+        oldEntityNum = oldSnapshot->entities[0].number;
+    }
+
+    // Likewise when we don't have an old snapshot, oldEntities just has to be an empty
+    // vector so that we skip step (4)
+    std::vector<entityState_t> dummyEntities;
+    auto& oldEntities = oldSnapshot? oldSnapshot->entities : dummyEntities;
+    auto& newEntities = newSnapshot->entities;
+
+    unsigned int numEntities = MSG_ReadShort(msg);
+    newEntities.reserve(numEntities);
+
+	unsigned oldIndex = 0;
+
+    while (true) {
+        unsigned int newEntityNum = MSG_ReadBits(msg, GENTITYNUM_BITS);
+
+        if (newEntityNum == MAX_GENTITIES - 1) {
+            break;
+        }
 
-			continue;
-		}
+        if (msg->readcount > msg->cursize) {
+            Sys::Drop("CL_ParsePacketEntities: end of message");
+        }
 
-		if ( oldnum > newnum )
-		{
-			// delta from baseline
-			if ( cl_shownet->integer == 3 )
-			{
-				Com_Printf( "%3i:  baseline: %i\n", msg->readcount, newnum );
-			}
+        // (1) all entities that weren't specified between the previous newEntityNum and
+        // the current one are unchanged and just copied over.
+        while (oldEntityNum < newEntityNum) {
+            newEntities.push_back(oldEntities[oldIndex]);
```

Comment:
It is very unlikely to have a completely empty snapshot (given that the client itself is an entity) but it is better to not assume that. Will fix.